org.adempiere.base/src/org/compiere/model/MSequence.java

```diff
@@ -542,28 +574,64 @@ public static String getDocumentNoFromSeq(MSequence seq, String trxName, PO po)
 
 		//	create DocumentNo
 		StringBuilder doc = new StringBuilder();
-		if (prefix != null && prefix.length() > 0) {
-			String prefixValue = Env.parseVariable(prefix, po, trxName, false);
-			if (!Util.isEmpty(prefixValue))
-				doc.append(prefixValue);
-		}
+		if (!Util.isEmpty(prefixValue, true))
+			doc.append(prefixValue);
 
 		if (decimalPattern != null && decimalPattern.length() > 0)
 			doc.append(new DecimalFormat(decimalPattern).format(next));
 		else
 			doc.append(next);
 
-		if (suffix != null && suffix.length() > 0) {
-			String suffixValue = Env.parseVariable(suffix, po, trxName, false);
-			if (!Util.isEmpty(suffixValue))
-				doc.append(suffixValue);
-		}
-
+		if (!Util.isEmpty(suffixValue, true))
+			doc.append(suffixValue);
+		
 		String documentNo = doc.toString();
 		if (s_log.isLoggable(Level.FINER)) s_log.finer (documentNo + " (" + incrementNo + ")"
 				+ " - Sequence=" + AD_Sequence_ID + " [" + trx + "]");
 		return documentNo;
 	}
+	
+	private AtomicBoolean isSequenceNoLevel = null;
+	private AtomicBoolean isUsePrefixAsKey = null;
+	private AtomicBoolean isUseSuffixAsKey = null;
```

Comment:
This wouldn't make it thread safe either. If you must do lazy initialization and cache the result, use either:
1. `AtomicReference<Boolean> isSequenceNoLevel = new AtomicReference<>()` OR
2. Use primitive int so that you can have tri-state of the variable. 
`private int isSequenceNoLevel=2; //2=unknown, 1=true, 0=false`

Reply:
Hi, I guess it's not worth the effort, the operation is not that expensive, I would just remove the global variables and call the `getPrefix().contains(KEY_CONTEXT_VARIABLE+"@")` each time, wdyt?

Reply:
done using AtomicReference<Boolean>